packages/flutter/lib/src/gestures/resampler.dart

```diff
@@ -335,3 +354,54 @@ class PointerEventResampler {
   /// Returns `true` if pointer is currently down.
   bool get isDown => _isDown;
 }
+
+/// Resamples pointer events from multiple devices.
+class MultiDeviceResampler {
+  // Resamplers used to filter incoming pointer events.
+  final Map<int, PointerEventResampler> _resamplers = <int, PointerEventResampler>{};
+
+  /// Enqueue `event` for resampling.
+  void addEvent(PointerEvent event) {
+    final PointerEventResampler resampler = _resamplers.putIfAbsent(
+      event.device,
+      () => PointerEventResampler(),
+    );
+    resampler.addEvent(event);
+  }
+
+  /// Resample enqueued events according to the given `sampleTime` and dispatch
+  /// results through `handlePointerEvent`.
+  ///
+  /// The `nextSampleTime` is the timestamp of the next frame. A positive value
+  /// allows early processing of up and removed events. This improves
+  /// resampling of these events, which is important for fling animations.
+  void sample(Duration sampleTime, Duration nextSampleTime, HandleEventCallback handlePointerEvent) {
+    // Iterate over active resamplers and sample pointer events for
+    // current sample time.
+    final List<BatchEventBuilder> batchBuilder = <BatchEventBuilder>[];
+    for (final PointerEventResampler resampler in _resamplers.values) {
+      batchBuilder.addAll(resampler.sample(sampleTime, nextSampleTime));
```

Comment:
While reading this I have been wondering why we need all this indirection via the `BatchEventBuilder`. For my own understanding: Is it possible that `resampler.sample` here my return an empty list and the `endOfBatch` event is actually produced by a Resampler other then the last one in `_resamplers.values`? If that's not true, could we just tell the last `resampler.sample` method that it is responsible for producing an `endOfBatch` event?

Reply:
If we have to sample all resamplers before we can determine the endOfBatch event, could this be simplified by just having resampler.sample return a list of pointer events and before calling handlePointerEvent below we copy the last event of the collected list to set `endOfBatch: true`?